lib/AST/ClangTypeConverter.cpp

```diff
@@ -0,0 +1,691 @@
+  // map the underlying types of those typealiases back to the builtin
+  // type.  These typealiases frequently create a many-to-one mapping,
+  // so just use the first type that mapped to a particular underlying
+  // type.
+  //
+  // This is the last thing that happens before asserting that the
+  // struct type doesn't have a mapping.  Furthermore, all of the
+  // builtin types are pre-built in the clang ASTContext.  So it's not
+  // really a significant performance problem to just cache all them
+  // right here; it makes making a few more entries in the cache than
+  // we really need, but it also means we won't end up repeating these
+  // stdlib lookups multiple times, and we have to perform multiple
+  // lookups anyway because the MAP_BUILTIN_TYPE database uses
+  // typealias names (like 'CInt') that aren't obviously associated
+  // with the underlying C library type.
+
+  auto stdlib = Context.getStdlibModule();
+  assert(stdlib && "translating stdlib type to C without stdlib module?");
+  auto &ctx = ClangASTContext;
+
+  if (!StdlibTypesAreCached) {
+    auto cacheStdlibType = [&](StringRef swiftName,
+                               clang::BuiltinType::Kind builtinKind) {
+      Type swiftType = getNamedSwiftType(stdlib, swiftName);
+      if (!swiftType) return;
+
+      auto &sema = Context.getClangModuleLoader()->getClangSema();
+      // Handle Int and UInt specially. On Apple platforms, these correspond to
+      // the NSInteger and NSUInteger typedefs, so map them back to those typedefs
+      // if they're available, to ensure we get consistent ObjC @encode strings.
+      if (swiftType->getAnyNominal() == Context.getIntDecl()) {
+        auto NSIntegerTy = getClangBuiltinTypeFromTypedef(sema, "NSInteger");
+        if (!NSIntegerTy.isNull()) {
+          Cache.insert({swiftType->getCanonicalType(), NSIntegerTy});
+          return;
+        }
+      } else if (swiftType->getAnyNominal() == Context.getUIntDecl()) {
+        auto NSUIntegerTy = getClangBuiltinTypeFromTypedef(sema, "NSUInteger");
+        if (!NSUIntegerTy.isNull()) {
+          Cache.insert({swiftType->getCanonicalType(), NSUIntegerTy});
+          return;
+        }
+      }
+
+      // For something like `typealias CInt = Int32`, reverseBuiltinTypeMapping
+      // will get Int32 as the input, so we need to record the desugared type.
+      Cache.insert({swiftType->getCanonicalType(),
+                    getClangBuiltinTypeFromKind(ctx, builtinKind)});
+    };
+
+#define MAP_BUILTIN_TYPE(CLANG_BUILTIN_KIND, SWIFT_TYPE_NAME)          \
+    cacheStdlibType(#SWIFT_TYPE_NAME, clang::BuiltinType::CLANG_BUILTIN_KIND);
+#include "swift/ClangImporter/BuiltinMappedTypes.def"
+
+    // On 64-bit Windows, no C type is imported as an Int or UInt; CLong is
+    // imported as an Int32 and CLongLong as an Int64. Therefore, manually
+    // add mappings to C for Int and UInt.
+    if (Triple.isOSWindows() && Triple.isArch64Bit()) {
```

Comment:
This needs to be changed to match the behavior in this recent PR. https://github.com/apple/swift/pull/20474